### PR TITLE
Use map for internal lock unit statuses.

### PR DESCRIPTION
**Pull request by ExternalReality**
Use a map instead of an array to lock units.

### CODE REVIEW

**Comment by ExternalReality** [comment]
[Index subordinate unit docs by Unit Name. (Basically use a Map)](https://trello.com/c/AUcuVRFJ/469-index-subordinate-unit-docs-by-unit-name-basically-use-a-map)

File: state/machine.go [hunk 1]
```diff
@@ -222,7 +222,7 @@ type upgradeSeriesLockDoc struct {
 	ToSeries      string                    `bson:"to-series"`
 	FromSeries    string                    `bson:"from-series"`
 	MachineStatus model.UpgradeSeriesStatus `bson:"machine-status"`
-	UnitStatuses  []unitStatus              `bson:"unit-statuses"`
+	UnitStatuses  map[string]unitStatus     `bson:"unit-statuses"`
```
[the remark this answers is not in the document]
Comment (manadart) [reply state/machine.go]: Most of the reasoning is on the basis of the tag, but I think name is OK (this is the same as ID, right?) - it is consistent with other ID keys in the structure.

But do we need ID in `unitStatus` then?

File: state/machine.go [hunk 2]
```diff
@@ -2138,11 +2137,11 @@ func (m *Machine) unitsHaveChanged(unitNames []string) (bool, error) {
 }
 
 func (m *Machine) prepareUpgradeSeriesLock(unitNames []string, toSeries string) *upgradeSeriesLockDoc {
-	prepareUnits := make([]unitStatus, len(unitNames))
-	completeUnits := make([]unitStatus, len(unitNames))
-	for i, name := range unitNames {
-		prepareUnits[i] = unitStatus{Id: name, Status: model.PrepareStarted, Timestamp: bson.Now()}
-		completeUnits[i] = unitStatus{Id: name, Status: model.NotStarted, Timestamp: bson.Now()}
+	prepareUnits := make(map[string]unitStatus, len(unitNames))
+	completeUnits := make(map[string]unitStatus, len(unitNames))
```
Comment (manadart) [inline state/machine.go]: Might as well ditch this and rename `prepareUnits` to `unitStatuses`.

File: state/machine.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ExternalReality) [reply state/machine.go]: Right now it would appear that we don't need it. I will remove it.

**Comment by ExternalReality** [comment]
$$MERGE$$

**Pull request merged**